substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/PreMainSupport.java

```diff
@@ -0,0 +1,234 @@
+ * It is the agent developers' responsibility to implement a native version of their agent premain
+ * method. It can be implemented in two ways:
+ * <ul>
+ * <li>Isolate code by checking current runtime. For example: <code>
+ *         <pre>
+ *         String vm = System.getProperty("java.vm.name");
+ *         if ("Substrate VM".equals(vm)) {
+ *           // native image runtime
+ *         } else{
+ *           // JVM runtime
+ *         }
+ *         </pre>
+ *     </code></li>
+ * <li>Use {@link com.oracle.svm.core.annotate.TargetClass} API to implement a native image version
+ * premain.</li>
+ * </ul>
+ * </p>
+ */
+public class PreMainSupport {
+
+    private static final String PREMAIN_OPTION_PREFIX = "-XX-premain:";
+
+    class PremainMethod {
+        String className;  // full qualified premain class name
+        Method method;
+        Object[] args;
+
+        PremainMethod(String className, Method method, Object[] args) {
+            this.className = className;
+            this.method = method;
+            this.args = args;
+        }
+    }
+
+    private Map<String, String> premainOptions = new HashMap<>();
+    // Order matters
+    private List<PremainMethod> premainMethods = new ArrayList<>();
+
+    @Platforms({Platform.HOSTED_ONLY.class})
+    public void registerPremainMethod(String className, Method executable, Object... args) {
+        premainMethods.add(new PremainMethod(className, executable, args));
+    }
+
+    /**
+     * Retrieve premain options from input args.
+     * Keep premain options and return the rest args as main args.
+     * The premain options format:
+     * <br>
+     * -XX-premain:[full.qualified.premain.class]:[premain options]
+     * <br>
+     *
+     * @param args
+     * @return
+     */
+    public String[] retrievePremainArgs(String[] args) {
+        List<String> mainArgs = new ArrayList<>();
+
+        for (String arg : args) {
+            if (arg.startsWith(PREMAIN_OPTION_PREFIX)) {
+                String premainOptionKeyValue = arg.substring(PREMAIN_OPTION_PREFIX.length());
+                String[] pair = premainOptionKeyValue.split(":");
+                if (pair.length == 2) {
+                    premainOptions.put(pair[0], pair[1]);
+                }
+            } else {
+                mainArgs.add(arg);
+            }
+        }
+        return mainArgs.toArray(new String[0]);
+    }
+
+    public void invokePremain() {
+        for (PremainMethod premainMethod : premainMethods) {
+            try {
+                Object[] args = premainMethod.args;
+                if (premainOptions.containsKey(premainMethod.className)) {
+                    args[0] = premainOptions.get(premainMethod.className);
+                }
+                // premain method must be static
+                premainMethod.method.invoke(null, args);
+            } catch (Throwable t) {
+                VMError.shouldNotReachHere("Fail to execute " + premainMethod.className + ".premain", t);
```

Comment:
This should be a user-facing error, right? User code can throw, so I don't see this as something that is a `VMError`. What does the JVM do in this case?

Reply:
In JVM, if the `premain` method fails, the JVM starting process shall be terminated. See https://github.com/openjdk/jdk21u/blob/2971cb5769121b47ac8c8db1078d67680a19341f/src/java.instrument/share/native/libinstrument/InvocationAdapter.c#L623-L629

I use `VMError` because it's a fatal error during VM starts up. But the cause is indeed an user error.